### PR TITLE
Surface related memories in wake-up and fix search ranking

### PR DESCRIPTION
**Pull request by Iron-Ham**
## Summary

- Fixes `lore-search` in two ways. The previous code passed a `sort` parameter to Notion's search endpoint, which switches ranking from relevance to recency and demotes the query to a lexical filter. More critically, the parent-type filter only accepted `database_id` parents; Notion SDK v5 returns `data_source_id` parents in the wild, silently discarding every real result. Both are now fixed — search returns relevance-ranked memories and doesn't drop them on the floor.
- Wake-up now surfaces "Related to Open Loops" — memories relevance-matched against the union of open-loop subjects and objects via `MemoryService.search`. Deduped against the digest and Recent Memories, capped at 5, gated on the presence of open loops.
- `MemoryService.search` accepts `includeContent: false` so the shell hook's related-memories pass avoids N+1 markdown round-trips.
- MCP wake-up tool and shell hook now share one aggregate loader (`loadWakeUpData`) that owns memories, facts, decisions, and related memories. Hook opts out of decisions via `includeDecisions: false` to keep session-start latency tight.

## Why

`lore-search` was silently broken in production. An agent that tried to recall a memory by entity name — e.g. `"Historic autolabel tail amplification"` — got zero results and fell back to raw Notion search. Root cause was two independent bugs that only surfaced together: (1) `sort` demoting semantic search to recency-filtered lexical matching, and (2) the parent-type filter rejecting every SDK-v5 page because they have `parent.type === "data_source_id"`, not `"database_id"`. Unit tests didn't catch the parent-type issue because they used synthetic `database_id` fixtures; only a real smoke test against a live vault exposed it.

`lore-wake-up` was topic-blind. It surfaced the most recently edited memories regardless of what mattered. But the signal for what matters is already in the vault — as open loops (tracking-predicate facts like `blocked_by` / `needs_action` / `waiting_on`). Walking those back to the memories that explain them turns wake-up from "here's recent stuff" into "here's recent stuff plus the memories that explain your outstanding work."

## What changed

- `MemoryService.search()` — drop `sort`, set `page_size: 100`, trim results to caller's `limit` before the markdown round-trip. Parent-type filter accepts both `database_id` and `data_source_id`. New `includeContent: false` option skips the per-page markdown fetch.
- `loadWakeUpData` — absorb the decision queries previously duplicated in the MCP tool, extract entities from open loops (case-insensitive dedupe, drop sub-3-char fragments, cap at 10 candidates), join into one relevance query, fetch via `memories.search`, dedupe against already-surfaced, cap. New `WakeUpData` fields: `proposedDecisions`, `overdueDecisions`, `relatedMemories`. New opts: `relatedMemoryLimit`, `includeDecisions`.
- MCP `lore-wake-up` — render "Related to Open Loops" between Recent Memories and Decisions Requiring Attention.
- Shell wake-up hook — render the same section in a compact format; pass `includeDecisions: false` and `includeMemoryContent: false` to keep the hot path fast.
- Docs: `README.md`, `src/core/AGENTS.md` updated.

## Design notes

- **Why semantic search over `title.contains`**: real-world open-loop subjects are phrase-shaped (`"PR #25650 outlook label.applied classifier"`), not bare entity names. Substring-on-title misses the memory that explains them (`"PR #25650 label.applied classifier: false positives on any updated categorized message"`) because the full phrase isn't a verbatim substring. Notion's vector index scores title AND body against the union query and surfaces the right memory. This was validated empirically against the Mail repo's vault before landing.
- **Fetch-window scaling**: `Math.min(NOTION_PAGE_SIZE, relatedLimit + alreadySurfaced.size)`. Guarantees at least `relatedLimit` survivors under dedupe unless Notion itself returns fewer candidates.
- **Hook perf**: `includeDecisions: false` skips two Notion round-trips (proposed + overdue decisions) on the hot path, since the hook doesn't render decisions. `includeMemoryContent: false` skips N+1 `retrieveMarkdown` calls for recent and related memories.
- **Known characteristic**: when recent work *heavily* overlaps with open-loop topics, the top relevance hits for the seed query will already be in Recent Memories, and dedupe will push the Related section to positions 6+ — which can look tangential. Not a bug; a design characteristic of relevance-ranking + dedupe. Documented in the rendered section copy.

## Test plan

- [x] `npm run typecheck`
- [x] `npm run lint`
- [x] `npm run test` — 126 tests, 14 files
- [x] `npm run build`
- [x] **Live smoke on this repo's vault**: `lore search "wakeup"` returns three relevance-ranked memories (previously zero). Parent-type fix validated by direct `client.search()` inspection showing `parent.type === "data_source_id"` for real pages.
- [x] **Live smoke on Mail repo's vault**: the original failing query — `lore search "Historic autolabel tail amplification"` — now returns the correct memory on first try (previously zero). Related-memories pass on Mail returns 5 memories relevance-matched against 5 open loops; dedupe against Recent Memories correctly filters out hits that are already on screen.
- [x] **Hook smoke on Mail**: renders clean output with Project header, Recent Memories (date-bucketed), Open Loops, Related to Open Loops, and Active Facts sections. No decisions sections (confirming `includeDecisions: false` takes effect). Hook wall time ~2.7s total (Node + Notion round-trips).
- [x] **Loader latency on Mail** (156 memories, 133 facts): MCP shape (with content + decisions) ~2.8s; hook shape (no content, no decisions) ~1.1s.

### CODE REVIEW

File: src/mcp/tools/context.ts [hunk 1]
```diff
@@ -86,26 +86,25 @@ export function registerContextTools(server: McpServer, services: LoreServices):
...
+          knowledgeFacts,
+          proposedDecisions,
+          overdueDecisions,
+          relatedMemories,
+        } = await loadWakeUpData(services, {
+          projectId: projectId ?? undefined,
+          memoryLimit: limit,
+          // Honor the caller's explicit limit even when a digest is present:
+          // the trim is a default, not a cap the user can't override.
+          memoryLimitWithDigest: limit,
```
Comment (Iron-Ham) [inline src/mcp/tools/context.ts:108]: The new related-memory section makes the `limit` option stop being a true upper bound on returned memories. We pass the caller's limit through to `memoryLimit`/`memoryLimitWithDigest`, but `relatedMemoryLimit` stays at the default 5, so a `limit: 1` wake-up can still append five full related-memory entries. Callers use this knob to stay within prompt budgets, so this changes the tool contract in a surprising way. Can we either thread `limit` through to `relatedMemoryLimit` as well, or explicitly rename/reword the option so it no longer promises a hard cap?

File: src/core/wakeup.ts [hunk 2]
```diff
@@ -87,32 +131,40 @@ export async function loadWakeUpData(
...
+            limit: 1,
+            includeUnscoped: false,
+            sortBy: "created_time",
+          })
+        : Promise.resolve([]),
+      projectId
+        ? services.facts.queryBySubject("", { projectId })
+        : Promise.resolve([]),
+      projectId
+        ? services.decisions.list({ projectId, status: "proposed", limit: 20 })
```
Comment (Iron-Ham) [inline src/core/wakeup.ts:185]: This helper now always issues `decisions.list()` and `queryOverdue()` whenever `projectId` is present, but the shell hook caller only reads `digest`, `memories`, `openLoops`, `knowledgeFacts`, and `relatedMemories` from the result. Since the shell wake-up runs on every session start, that means we add two extra Notion round-trips to the hot path without changing the rendered output. Could we gate decision fetching behind an option, or render the decision sections in the hook too, so startup latency doesn't regress?

**Review by Iron-Ham — COMMENTED** [review]
I pulled the PR locally and ran `npm run test`, `npm run typecheck`, and `npm run lint`; those are green. I left two inline comments for regressions that still look worth fixing before merge: the wake-up `limit` no longer bounds total memory output, and the shared loader now adds two unused decision queries to the shell hook's session-start path.

File: src/mcp/tools/context.ts (hunk 1 above)
Comment (Iron-Ham) [reply src/mcp/tools/context.ts:108]: Good catch — threaded `limit` into `relatedMemoryLimit` as well so it acts as a per-section cap across both the recent and related buckets. A `limit: 1` wake-up now caps each section at 1 memory. Updated the Zod `describe` to document it as a per-section cap so future callers don't make the same assumption. Pushed as e00b7a9.

File: src/core/wakeup.ts (hunk 2 above)
Comment (Iron-Ham) [reply src/core/wakeup.ts:185]: Added an `includeDecisions` option (default `true` to preserve MCP behavior) and the hook now passes `includeDecisions: false`. Hook wake-up skips both `decisions.list` and `queryOverdue` — the two round-trips are gone from the session-start hot path. New test in `wakeup.test.ts` pins the opt-out. Pushed as e00b7a9.

File: src/core/wakeup.ts [hunk 3]
```diff
@@ -132,7 +192,62 @@ export async function loadWakeUpData(
...
+  for (const mem of memories) alreadySurfaced.add(mem.id)
+
+  let relatedMemories: Memory[] = []
+  if (projectId && openLoops.length > 0 && relatedLimit > 0) {
+    const entities = extractEntities(openLoops)
+    if (entities.length > 0) {
+      const candidates = await services.memories.list({
+        projectId,
+        titleAny: entities,
+        limit: RELATED_MEMORY_FETCH_LIMIT,
```
Comment (Iron-Ham) [inline src/core/wakeup.ts]: `lore-wake-up` now forwards the caller's `limit` into `relatedMemoryLimit`, but this fetch window is still hard-capped at 20 before dedupe. That means a caller asking for `limit: 50` can never actually get 50 related memories, and even the default 5 can be starved if the first 20 hits are mostly duplicates of the digest/recent lists we drop below. It would be safer to make the candidate window scale with `relatedLimit` (or keep paging until we collect enough non-duplicate results), and add a test that exercises the duplicate-heavy case.

**Review by Iron-Ham — COMMENTED** [review]
Re-reviewed the force-pushed update on `e00b7a9`. The two earlier concerns are fixed: `includeDecisions: false` removes the extra hook-path queries, and the MCP tool now threads the caller's `limit` into `relatedMemoryLimit`. I ran `npm run test`, `npm run typecheck`, and `npm run lint`; those are still green. I left one follow-up inline comment for a remaining edge case in the related-memory fetch window.

File: src/core/wakeup.ts (hunk 3 above)
Comment (Iron-Ham) [reply src/core/wakeup.ts]: Good follow-up — replaced the hard-20 cap with a formula that scales with both the caller's ask and the already-surfaced set: `fetchLimit = min(NOTION_PAGE_SIZE, relatedLimit + alreadySurfaced.size)`. That guarantees `relatedLimit` survivors in every case where Notion can fit the request in one query — including the duplicate-heavy case where every candidate collides with the digest or recents.

Added two tests in `wakeup.test.ts`:
- `scales the related-memory fetch window with relatedLimit + already-surfaced size` — pins the fetch size for a `relatedMemoryLimit: 50` call.
- `does not starve related memories when candidates are mostly duplicates` — 4 of 7 candidates are dupes of digest/recents, verifies all 3 fresh survivors come through.

Skipped full pagination: `MemoryService.list` doesn't currently expose a cursor, and the only unreachable case is a `limit: 50` call where all 100 fetched candidates also happen to share entity titles with the 51 already-surfaced memories. Happy to add pagination in a follow-up if that edge case shows up in practice. Pushed as 5694aa2.

**Review by Iron-Ham — COMMENTED** [review]
Re-reviewed the force-pushed update on `5694aa2`. The fetch-window concern is addressed: related-memory candidate loading now scales with `relatedMemoryLimit + alreadySurfaced.size` and is covered by duplicate-heavy tests. I reran `npm run test`, `npm run typecheck`, and `npm run lint`; all passed. No new findings from this pass.

**Review by Iron-Ham — COMMENTED** [review]
Re-reviewed the force-pushed update on `1420c61` together with the refreshed PR description. The functional changes line up with the new write-up: `MemoryService.search()` now handles both `database_id` and `data_source_id` parents, related memories are loaded through semantic search, and the hook still opts out of decision queries on the hot path. I reran `npm run test`, `npm run typecheck`, `npm run lint`, and `npm run build`; all passed. No new code findings from this pass. One small non-blocking follow-up: README and a couple of wake-up comments still describe the related section as title/entity-name matching, while the implementation now uses semantic search across titles and bodies.
